src/common/com/intellij/plugins/haxe/util/HaxeResolveUtil.java

```diff
@@ -592,37 +593,37 @@ private static HaxeClass tryResolveClassByQNameWhenGetQNameFail(@NotNull PsiElem
     }
 
     String className = type.getText();
-    HaxeClass result;
+    PsiElement result = null;
 
     if (className != null && className.indexOf('.') == -1) {
       final HaxeFileModel fileModel = HaxeFileModel.fromElement(type);
-
-      result = searchInSameFile(fileModel, className);
-      if (result == null) {
-        PsiElement psiElement = searchInImports(fileModel, className);
-
-        if (psiElement != null && psiElement instanceof HaxeClass) {
-          result = (HaxeClass)psiElement;
-        }
+      if (fileModel != null) {
+        result = searchInSameFile(fileModel, className);
+        if (result == null) result = searchInImports(fileModel, className);
+        if (result == null) result = searchInSamePackage(fileModel, className);
       }
-      if (result == null) result = searchInSamePackage(fileModel, className);
     } else {
       className = tryResolveFullyQualifiedHaxeReferenceExpression(type);
       result = findClassByQName(className, type.getContext());
     }
 
     result = result != null ? result : findClassByQName(className, type.getContext());
 
-    return result;
+    return result instanceof HaxeClass ? (HaxeClass)result : null;
   }
 
   @Nullable
-  public static HaxeClass searchInSameFile(HaxeFileModel file, String name) {
-    final HaxeClassModel classModel = file.getClassModel(name);
-    if (classModel != null) {
-      return classModel.haxeClass;
-    }
-    return null;
+  public static PsiElement searchInSameFile(@NotNull HaxeFileModel file, @NotNull String name) {
+    List<HaxeClassModel> models = file.getClassModels();
+    final HaxeModel result = Stream.concat(
+      models.stream().filter(model -> name.equals(model.getName())),
+      models.stream().filter(HaxeClassModel::isEnum)
+        .map(model -> model.getField(name))
+        .filter(Objects::nonNull))
```

Comment:
This is not a criticism; it's a realization of my own frailty:  I hate that I have to read this kind of stuff four times before I get it.  It's not straightforward to me, so I fear this style of code is going to lead to maintenance headaches -- but I may be wrong!  Maybe I'll just get used to it.

I wonder if there is a way to structure this code better?
```java
final Stream modelStream = models.stream().filter(model -> name.equals(model.getName());
final Stream enumStream = models.stream().filter(HaxeClassModel::isEnum)
                           .map(model -> model.getField(name))
                           .filter(Objects::nonNull));
final HaxeModel result = Stream.concat(modelStream, enumStream)
                            .findFirst()
                            .orElse(null);
```
What do you think?  Should we strive for this style when we have multiple streams interacting?

Reply:
It has sense, readability in such cases is getting worse. Your code looks nicer.